### DISCUSSION TITLE
Developer Setup on Pi

### DISCUSSION THREAD

**Discussion by vszander** [micro-nova/AmpliPi#91] [Q&A]
Thank you,  Jason; Lincoln; Mark   - this is a great endeavor !  

 I started watching a couple of weeks ago and missed out on the Developer Kickstarter - so I contributed at the next level.   I'm a hack SBC developer - but I'd like to contribute.   My Vantec NBA-200U  was delivered yesterday and my  pi  is playing XM radio over it now over the 'front' channel.  If it will be possible to get an 'early adopter'  product,  I'd like to help with some home automation ideas I have.

I followed the DEVELOPING.md  instructions an did a deploy.  The first time through:  I followed the script's instructions to " set RPI_IP_ADDRESS.    When I  run_debug_webserver --mock-ctrl       I get 3 errors:   
line 41:  venv/bin/activate  no such file   (  I have venv installed and a django instance installed)
line 43:  flask:  command  not found  (  I've not installed flask )
line 44:  deactivate (command not found)  

I've attached (renamed to .txt) 
[asound.txt](https://github.com/micro-nova/AmpliPi/files/6143124/asound.txt)
my asound.conf   for the mock board.   did I follow instructions correctly ?

Please advise on how I should correctly setup for development...

**Comment by linknum23** [comment]
Looks like there is a bug in the deploy script that doesn't handle the case where venv is installed but hasn't been created. Let me see if i can give you a quick fix.

**Reply by linknum23** [reply]
Can you attach the output of the deploy script? It calls the `update_config.bash` script that installs most of the necessary packages. You can also try running update_config.bash on its own and checking the output.

**Reply by linknum23** [reply]
You may also want to ad the line `set -e` to the top of the `update_config.bash` script so that it stops at the first error.

**Comment by vszander** [comment]
pi@raspberrypi:~/amplipi/AmpliPi/scripts $ ./deploy

Please set RPI_IP_ADDRESS, for example:
export RPI_IP_ADDRESS=192.168.0.140
for ssh key access
export RPI_IP_ADDRESS=pi@192.168.0.140

**Reply by vszander** [reply]
Thanks, Lincoln.
   your suggestion to just run the update_config.bash seems to have worked.  It ran through to completion but had some errors:
E: Unable to locate package /home/pi/amplipi/AmpliPi/config/../debs
E: Couldn't find any package by regex '/home/pi/amplipi/AmpliPi/config/../debs'
E: Unable to locate package /home/pi/amplipi/AmpliPi/config/../debs
E: Couldn't find any package by regex '/home/pi/amplipi/AmpliPi/config/../debs'
Updating AmpliPi's Nginx Unit webserver configuration
Failed to restart unit.service: Unit unit.service failed to load: No such file or directory.

when I tried to http to the PI from port 80, my Django page opened (which I was actually happy to see)... can I set nginx to serve AmpliPi on a different port ?

**Reply by linknum23** [reply]
~~At the current time most if not all of the scripts and some of the code assume they are being installed to the pi users home directory, `/home/pi`. It looks like you have modified some of the scripts to install to `/home/pi/AmpliPi`. In the future we could make this home/installation folder a configuration parameter, for now is there anyway you can use the '/home/pi' directory?~~

I understand now. You've checked out AmpliPi on the pi! The standard development setup is on a Linux computer. The deploy script is made for copying the files from my development computer to the pi. Here's what you need to do to make this work for your setup:

First of all you will need to disable the Nginx Unit webserver, comment out these lines in config/update_config.bash similar to below. This removes the webserver that will conflict with your Django setup:
```
# install nginx unit from debians built on the pi and configure its service
#unit_installed=$(sudo apt list --installed 2> /dev/null | grep unit-python -c)
#if [ 0 -eq "${unit_installed}" ]; then
#  echo "installing unit"
#  sudo apt update && sudo apt install -y ${SCRIPT_DIR}/../debs/unit_*.deb ${SCRIPT_DIR}/../debs/unit-#python3.7_*.deb
#else
#  echo "unit already installed"
#fi
#bash ${SCRIPT_DIR}/update_web.bash
```

Now reconfigure your setup to copy locally. We need to move the amplipi folder, because the deploy script will overwrite it.

```
mv ~/amplipi/AmpliPi ~ # move the AmpliPi repository to the root of the pi users directory ( you will only need to do this once)
export RPI_IP_ADDRESS=localhost # tell the deploy script to copy files locally
./AmpliPi/scripts/deploy # actually copy the files and configure the pi
```

As for the Django conflict, feel free to run the debug webserver using `./scripts/run_debug_webserver`, this runs on port 5000. It looks like I will need to update the DEVELOPING.md, this script was recently renamed from `scripts/run_webserver`.

**Reply by vszander** [reply]
Thanks again.   I'll blow this install away; do a fresh git pull into the /home/pi  directory  and re-run.

**Reply by linknum23** [reply]
@vszander I've updated the comment above. If you checkout ampli in /home/pi you wont need the mv command.

**Reply by vszander** [reply]
my Django is served by apache... so  I'm hoping they can both co-exist.  since nginx is serving for port 5000.   I didn't see you post about not installing 'unit'   so it got installed ( I can always roll-back by pi's  SD card ).

But my update_config.bash is throwing errors  ( python 3.4 is installed - maybe I need to get 3.7 ), maybe it wasn't designed to build from the pi itself ?
output:
E: Unable to locate package /home/pi/AmpliPi/config/../debs
E: Couldn't find any package by regex '/home/pi/AmpliPi/config/../debs'
E: Unable to locate package /home/pi/AmpliPi/config/../debs
E: Couldn't find any package by regex '/home/pi/AmpliPi/config/../debs'
Updating AmpliPi's Nginx Unit webserver configuration
Failed to restart unit.service: Unit unit.service failed to load: No such file or directory.

If it would help,  I could collaborate w/ you on webex ?  my gmail is:   daytona.zander1@gmail.com

**Reply by vszander** [reply]
my error from run_debug_webserver:
pi@raspberrypi:~/AmpliPi/scripts $ ./run_debug_webserver 
```
kill: usage: kill [-s sigspec | -n signum | -sigspec] pid | jobspec ... or kill -l [sigspec]
librespot: no process found
pianobar: no process found
shairport-sync: no process found
 * Serving Flask app "amplipi.app:create_app(False,False)"
 * Environment: development
 * Debug mode: on
Traceback (most recent call last):
  File "/home/pi/AmpliPi/venv/bin/flask", line 11, in <module>
    sys.exit(main())
...
...
 File "/home/pi/AmpliPi/venv/lib/python3.4/site-packages/deepdiff/__init__.py", line 10, in <module>
    from .diff import DeepDiff
  File "/home/pi/AmpliPi/venv/lib/python3.4/site-packages/deepdiff/diff.py", line 680
    **self.deephash_parameters,
                              ^
SyntaxError: invalid syntax
```

**Reply by linknum23** [reply]
The deploy script should have copied everything into `/home/pi/`. Running stuff from the repo folder `/home/pi/AmpliPi` will not work since the base folder is now `/home/pi/AmpliPi` instead of `/home/pi`.  Try running the debug webserver from there:
```
cd ~
./scripts/run_debug_webserver --mock-ctrl # --mock-ctrl pretends there is a AmpliPi controller there
```

**Reply by linknum23** [reply]
We are working on packaging AmpliPi into a release, so all of the relative path issues will be solved relatively soon. Sorry that you had so many issues with this.

**Comment by vszander** [comment]
Thanks,  Lincoln.
   I understand, now,  how you are developing -  and it makes sense.   I'm
working to create a similar setup.   My delay has been:  I had to upgrade
my version of python and openssl...   hopefully I'll have a breakthrough
soon.

I'm trying not to burn up your time ( which is better spent getting ready
for production)  on my silly questions !

Kind regards,
Van

**Comment by vszander** [comment]
Hi Lincoln,
   Hope you are well !    It looks like there are some delays in developer
units ?     I'd like to contribute -  but  I had so much trouble getting an
SD card image  to work ..

Is there a way I could get a linux  'dd'   image file for a  a basic/
latest  build ?  I purchased the recommended sound card and would like to
see if I can contribute.

Thank you for your efforts !


Van

**Reply by linknum23** [reply]
What pi are you using? In our office we have a pi3 and a pi4, so we can pretty easily make a working image for either a pi3 or pi4.

**Comment by vszander** [comment]
I also have both.   But it seems like the AmpliPi  will ship  with
Raspberry Pi3...  so lets go with that.
I think you also asked what my IP was...
   my public is:  67.191.140.146
inside,  my NAT'd  addresses.
192.168.35.138  Raspi 3
192.168.35.144  Raspi 4


If you plan to ship with Raspi4 ,  I'd also like to work on a Raspi4.

Thanks !


Van

**Reply by linknum23** [reply]
@micronova-jb is working on creating a defacto image for our first 5 AmpliPis. We should be able to get you something that works in the next couple of days.

**Comment by vszander** [comment]
I saw the update on the inaugural 5 !   congratulations!

if you get a minute,  I'd like to download the img    so I can contribute.

VZ

**Reply by micronova-jb** [reply]
Hey Van! Sorry for the delay - I was trying to figure out a way to actually get this to you today. Does this Google Drive link work for you? You may have to request access: https://drive.google.com/file/d/1HLGoI8innn-Ur6-T1AUBfP6ZJbJZ4Etn/view?usp=sharing

**Reply by micronova-jb** [reply]
Better yet, this should be easier: https://www.dropbox.com/s/hpqvukgqshzna7d/aptest_1.4.img.7z?dl=0
